posts/MarkoB/2017-02-06-validator-service-loader.adoc

```diff
@@ -0,0 +1,290 @@
+users of your lib only with the relevant classes and hide implementation details from them, you don't want to
+expose your validator implementation by mentioning it as a `validatedBy` parameter of `@Constraint(validatedBy = { })`
+annotation (as well as making an import of it from your `*.impl` or `*.internal` package where all the magic things
+are happening). It's a good practice to keep your public API clean and hide any implementation details. There's
+http://http://in.relation.to/2017/01/31/preventing-leaky-apis-with-jqassistant/[this] great post about a tool that
+can help you analyze your code and make sure that it is not leaking anything that is implementation only and
+not for the public.
+
+Let's get back to validating our beans - as that's why we all gathered here, right ? :)
+
+Possible version of our new constraint annotation '@MinDuration' might look like this:
+
+```java
+@Documented
+@Constraint(validatedBy = { })
+@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE })
+@Retention(RUNTIME)
+@Repeatable(List.class)
+@ReportAsSingleViolation
+public @interface DurationMin {
+
+	String message() default "{org.hibernate.validator.constraints.time.DurationMin.message}";
+
+	Class<?>[] groups() default { };
+
+	Class<? extends Payload>[] payload() default { };
+
+	long value() default 0;
+
+	ChronoUnit units() default ChronoUnit.NANOS;
+
+	/**
+	 * Defines several {@code @DurationMin} annotations on the same element.
+	 */
+	@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE })
+	@Retention(RUNTIME)
+	@Documented
+	@interface List {
+
+		DurationMin[] value();
+	}
+}
+```
+
+And we would place it in a public package of our lib - `org.hibernate.validator.constraints.time`. Now that we
+have an annotation we would want to start writing a validator for `Duration` and this newly created annotation.
+To do that you need to implement `ConstraintValidator<DurationMin, Duration>` interface, which contains two methods
+
+* `initialize` - initializes validator based on annotation parameters.
+* `isValid` - performs actual validation
+
+A possible implementation can be next:
+
+```java
+public class DurationMinValidator implements ConstraintValidator<DurationMin, Duration> {
+
+	private Duration duration;
+
+	@Override
+	public void initialize(DurationMin constraintAnnotation) {
+		this.duration = Duration.of( constraintAnnotation.value(), constraintAnnotation.units() );
+	}
+
+	@Override
+	public boolean isValid(Duration value, ConstraintValidatorContext context) {
+		// null values are valid
+		if ( value == null ) {
+			return true;
+		}
+		return duration.compareTo( value ) < 1;
+	}
+}
+```
+
+As was mentioned earlier - you want to hide this implementation from API users by putting it somewhere in internal
+packages, in our case it will be - `org.hibernate.validator.internal.constraintvalidators.hv.time`.
```

Comment:
:) It's a very serious module here! :D

yep, Gunnar mentioned that one can motivate usage of ServiceLoader lookup based on the fact that implementation is hidden from the API users. So that's why I put this sentence together, as well as the part about "... good practice to keep public API clean ..."

But I'll stop screaming about it :) if you say so.